packages/core/js-client/src/compilerSupport/conversions.ts

```diff
@@ -14,222 +14,153 @@
       return null;
-    })
-    .with({ tag: "option" }, (opt) => {
-      if (value === null || value === undefined) {
-        return [];
-      } else {
-        return [ts2aqua(value, opt.type)];
-      }
-    })
-    .with({ tag: "scalar" }, { tag: "bottomType" }, { tag: "topType" }, () => {
-      return value;
-    })
-    .with({ tag: "array" }, (arr) => {
-      assert(Array.isArray(value), "Should not be possible, bad types");
-      return value.map((y) => {
-        return ts2aqua(y, arr.type);
-      });
-    })
-    .with({ tag: "struct" }, (x) => {
-      return Object.entries(x.fields).reduce((agg, [key, type]) => {
-        const val = ts2aqua(value[key], type);
-        return { ...agg, [key]: val };
-      }, {});
-    })
-    .with({ tag: "labeledProduct" }, (x) => {
-      return Object.entries(x.fields).reduce((agg, [key, type]) => {
-        const val = ts2aqua(value[key], type);
-        return { ...agg, [key]: val };
-      }, {});
-    })
-    .with({ tag: "unlabeledProduct" }, (x) => {
-      return x.items.map((type, index) => {
-        return ts2aqua(value[index], type);
-      });
-    })
-    // uncomment to check that every pattern in matched
-    // .exhaustive()
-    .otherwise(() => {
-      throw new Error("Unexpected tag: " + jsonify(type));
+    } else {
+      return aqua2ts(value[0], schema.type);
+    }
+  } else if (
+    schema.tag === "scalar" ||
+    schema.tag === "bottomType" ||
+    schema.tag === "topType"
+  ) {
+    return value;
+  } else if (schema.tag === "array") {
+    if (!Array.isArray(value)) {
+      throw new Error("Value and schema doesn't match");
+    }
+
+    return value.map((y) => {
+      return aqua2ts(y, schema.type);
     });
+  } else if (schema.tag === "unlabeledProduct") {
+    if (!Array.isArray(value)) {
+      throw new Error("Value and schema doesn't match");
+    }
 
-  return res;
-};
-
-/**
- * Convert return type of the service from it's typescript representation to representation in aqua
- * @param returnValue - the value as represented in typescript
- * @param arrowType - the arrow type which describes the service
- * @returns - value represented in aqua
- */
-export const returnType2Aqua = (
-  returnValue: any,
-  arrowType: ArrowType<NonArrowType>,
-) => {
-  if (arrowType.codomain.tag === "nil") {
-    return {};
-  }
-
-  if (arrowType.codomain.items.length === 0) {
-    return {};
+    return value.map((y, i) => {
+      return aqua2ts(y, schema.items[i]);
+    });
+  } else if (["labeledProduct", "struct"].includes(schema.tag)) {
+    if (typeof value !== "object" || value == null || Array.isArray(value)) {
+      throw new Error("Value and schema doesn't match");
+    }
+
+    return Object.entries(schema.fields).reduce((agg, [key, type]) => {
+      const val = aqua2ts(value[key], type);
+      return { ...agg, [key]: val };
+    }, {});
+  } else {
+    throw new Error("Unexpected tag: " + JSON.stringify(schema));
   }
+}
+
+export function ts2aqua(
```

Comment:
this function needs doc and some refactoring. Too many conditions, errors, returns – super hard to understand intentions.